wgengine/wgtun/tun.go

```diff
@@ -0,0 +1,264 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package wgtun provides a TUN struct implementing the tun.Device interface
+// with additional features as required by wgengine.
+package wgtun
+
+import (
+	"errors"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/tailscale/wireguard-go/device"
+	"github.com/tailscale/wireguard-go/tun"
+	"tailscale.com/types/logger"
+	"tailscale.com/wgengine/filter"
+	"tailscale.com/wgengine/packet"
+)
+
+const (
+	readMaxSize   = device.MaxMessageSize
+	readOffset    = device.MessageTransportHeaderSize
+	MaxPacketSize = device.MaxContentSize
+)
+
+var (
+	ErrClosed       = errors.New("device closed")
+	ErrFiltered     = errors.New("packet dropped by filter")
+	ErrPacketTooBig = errors.New("packet too big")
+)
+
+// TUN wraps a tun.Device from wireguard-go,
+// augmenting it with filtering and packet injection.
+// All the added work happens in Read and Write:
+// the other methods delegate to the underlying tdev.
+type TUN struct {
+	logf logger.Logf
+	// tdev is the underlying TUN device.
+	tdev tun.Device
+
+	// buffer stores the oldest unconsumed packet from tdev.
+	// It is made a static buffer in order to avoid graticious allocation.
+	buffer [readMaxSize]byte
+	// bufferConsumed synchronizes access to buffer (shared by Read and poll).
+	bufferConsumed chan struct{}
+
+	// closed signals poll (by closing) when the device is closed.
+	closed chan struct{}
+	// errors is the error queue populated by poll.
+	errors chan error
+	// outbound is the queue by which packets leave the TUN device.
+	// The directions are relative to the network, not the device:
+	// inbound packets arrive via UDP and are written into the TUN device;
+	// outbound packets are read from the TUN device and sent out via UDP.
+	// This queue is needed because although inbound writes are synchronous,
+	// the other direction must wait on a Wireguard goroutine to poll it.
+	outbound chan []byte
+
+	// filterFlags control the verbosity of logging packet drops/accepts.
+	filterFlags filter.RunFlags
+	// filterMu locks the filter, since it may be in use
+	// during the netmap refresh when ipn/local updates it.
+	filterMu sync.Mutex
+	filter   *filter.Filter
+}
+
+func WrapTUN(logf logger.Logf, tdev tun.Device) *TUN {
+	tun := &TUN{
+		logf: logf,
+		tdev: tdev,
+		// bufferConsumed is conceptually a condition variable:
+		// a goroutine should not block when setting it, even with no listeners.
+		bufferConsumed: make(chan struct{}, 1),
+		closed:         make(chan struct{}),
+		errors:         make(chan error),
+		outbound:       make(chan []byte),
+		filterFlags:    filter.LogAccepts | filter.LogDrops,
+	}
+	go tun.poll()
+	// The buffer starts out consumed.
+	tun.bufferConsumed <- struct{}{}
+
+	return tun
+}
+
+func (t *TUN) Close() error {
+	// Other channels need not be closed: poll will exit gracefully after this.
+	close(t.closed)
```

Comment:
It's not a strict rule, but generally an `io.Closer` should do its best to not crash or do bad things if Close is called multiple times.